File: SlipieCoreKit/Sources/SlipieCoreKit/AudioEngine/ParameterMapper.swift
import Foundation

public final class ParameterMapper: Sendable {
    public init() {}

    public func map(biometrics: BiometricPacket, stage: SleepStage, soundscape: Soundscape) -> AudioParameters {
        let hrNorm = normalize(biometrics.heartRate, min: 40, max: 100)
        let hrvNorm = normalize(biometrics.hrv, min: 10, max: 100)

        let stageReverbBoost: Double
        let stageFilterMod: Double
        let stageTempoDamp: Double
        switch stage {
        case .awake:
            stageReverbBoost = 0.0; stageFilterMod = 0.0; stageTempoDamp = 0.0
        case .light:
            stageReverbBoost = 0.15; stageFilterMod = -0.1; stageTempoDamp = 0.1
        case .deep:
            stageReverbBoost = 0.45; stageFilterMod = -0.4; stageTempoDamp = 0.35
        case .rem:
            stageReverbBoost = 0.25; stageFilterMod = -0.15; stageTempoDamp = 0.2
        }

        let reverbWetness = clamp(Double(soundscape.baseParameters.reverbPreset) / 10.0 + stageReverbBoost + (hrNorm * 0.2), min: 0, max: 1)
        let filterCutoff = clamp(0.7 - (hrNorm * 0.3) + stageFilterMod + (hrvNorm * 0.1), min: 0.05, max: 1)
        let tempo = clamp(0.8 - (hrNorm * 0.4) - stageTempoDamp + (hrvNorm * 0.15), min: 0.05, max: 1.5)
        let volume = clamp(0.75 - (biometrics.motionIntensity * 0.2), min: 0.3, max: 1.0)

        return AudioParameters(
            volume: volume,
            tempo: tempo,
            filterCutoffNormalized: filterCutoff,
            reverbWetness: reverbWetness,
            oscillatorMix: soundscape.baseParameters.oscillatorMix,
            pitchShift: -(hrNorm * 4)
        )
    }

    private func normalize(_ value: Double, min: Double, max: Double) -> Double {
        clamp((value - min) / (max - min), min: 0, max: 1)
    }

    private func clamp(_ value: Double, min: Double, max: Double) -> Double {
        Swift.min(Swift.max(value, min), max)
    }
}

File: backend/app/audio/parameter_mapper.py
from app.audio.schemas import AudioParameters
from app.sessions.schemas import BiometricPacket, SleepStage
from app.soundscapes.schemas import Soundscape


def _clamp(value: float, min_val: float, max_val: float) -> float:
    return max(min_val, min(value, max_val))


def _normalize(value: float, min_val: float, max_val: float) -> float:
    return _clamp((value - min_val) / (max_val - min_val), 0.0, 1.0)


# Stage modifiers: (reverb_boost, filter_mod, tempo_damp)
_STAGE_MODS: dict[SleepStage, tuple[float, float, float]] = {
    SleepStage.awake: (0.0, 0.0, 0.0),
    SleepStage.light: (0.15, -0.1, 0.1),
    SleepStage.deep: (0.45, -0.4, 0.35),
    SleepStage.rem: (0.25, -0.15, 0.2),
}


def map_parameters(
    biometrics: BiometricPacket,
    stage: SleepStage,
    soundscape: Soundscape,
) -> AudioParameters:
    """Port of ParameterMapper.swift — must produce identical output."""

    hr_norm = _normalize(biometrics.heart_rate, 40, 100)
    hrv_norm = _normalize(biometrics.hrv, 10, 100)

    reverb_boost, filter_mod, tempo_damp = _STAGE_MODS[stage]

    reverb_wetness = _clamp(
        soundscape.base_parameters.reverb_preset / 10.0
        + reverb_boost
        + (hr_norm * 0.2),
        0, 1,
    )
    filter_cutoff = _clamp(
        0.7 - (hr_norm * 0.3) + filter_mod + (hrv_norm * 0.1),
        0.05, 1,
    )
    tempo = _clamp(
        0.8 - (hr_norm * 0.4) - tempo_damp + (hrv_norm * 0.15),
        0.05, 1.5,
    )
    volume = _clamp(0.75 - (biometrics.motion_intensity * 0.2), 0.3, 1.0)

    return AudioParameters(
        volume=volume,
        tempo=tempo,
        filter_cutoff_normalized=filter_cutoff,
        reverb_wetness=reverb_wetness,
        oscillator_mix=soundscape.base_parameters.oscillator_mix,
        pitch_shift=-(hr_norm * 4),
    )
